src/generate_conf.rs

```diff
@@ -0,0 +1,228 @@
+        let bridge_conn = fs::read_to_string(output_path.join("bridge0.nmconnection"))?;
+        let lo_conn = fs::read_to_string(output_path.join("lo.nmconnection"))?;
+
+        assert_eq!(exp_eth0_conn, eth0_conn);
+        assert_eq!(exp_bridge_conn, bridge_conn);
+        assert_eq!(exp_lo_conn, lo_conn);
+
+        // verify contents of the host mapping file
+        let mut exp_host_interfaces: Vec<HostInterfaces> = serde_yaml::from_str(
+            fs::read_to_string(exp_output_path.join(HOST_MAPPING_FILE))?.as_str(),
+        )?;
+        let mut host_interfaces: Vec<HostInterfaces> = serde_yaml::from_str(
+            fs::read_to_string(Path::new(out_dir).join(HOST_MAPPING_FILE))?.as_str(),
+        )?;
+
+        assert_eq!(exp_host_interfaces.len(), host_interfaces.len());
+
+        exp_host_interfaces.sort_by(|a, b| a.hostname.cmp(&b.hostname));
+        host_interfaces.sort_by(|a, b| a.hostname.cmp(&b.hostname));
+
+        for (x, y) in exp_host_interfaces
+            .iter_mut()
+            .zip(host_interfaces.iter_mut())
+        {
+            x.interfaces
+                .sort_by(|a, b| a.logical_name.cmp(&b.logical_name));
+            y.interfaces
+                .sort_by(|a, b| a.logical_name.cmp(&b.logical_name));
+        }
+
+        assert_eq!(exp_host_interfaces, host_interfaces);
+
+        // cleanup
+        fs::remove_dir_all(out_dir)?;
+
+        Ok(())
+    }
+
+    #[test]
+    fn generate_fails_due_to_missing_path() {
+        let error = generate("<missing>", "_out").unwrap_err();
+        assert_eq!(
+            error.to_string().contains("No such file or directory"),
+            true
+        )
+    }
+
+    #[test]
+    fn generate_config_fails_due_to_invalid_data() {
+        let err = generate_config("<invalid>".to_string()).unwrap_err();
+        assert_eq!(
+            err.to_string()
+                .contains("InvalidArgument: Invalid YAML string"),
+            true
+        )
+    }
+
+    #[test]
+    fn extract_interfaces_skips_loopback() -> Result<(), serde_yaml::Error> {
+        let net_state: nmstate::NetworkState = serde_yaml::from_str(
+            r#"---
+        interfaces:
+          - name: eth1
+            type: ethernet
+            mac-address: FE:C4:05:42:8B:AA
+          - name: bridge0
+            type: linux-bridge
+            mac-address: FE:C4:05:42:8B:AB
+          - name: lo
+            type: loopback
+            mac-address: 00:00:00:00:00:00            
+        "#,
+        )?;
+
+        let mut interfaces = extract_interfaces(&net_state);
+        assert_eq!(interfaces.len(), 2);
+
+        interfaces.sort_by(|a, b| a.logical_name.cmp(&b.logical_name));
+
+        let i1 = interfaces.get(0).unwrap();
+        assert_eq!(i1.logical_name, "bridge0");
+        assert_eq!(i1.mac_address, "FE:C4:05:42:8B:AB");
+
+        let i2 = interfaces.get(1).unwrap();
+        assert_eq!(i2.logical_name, "eth1");
+        assert_eq!(i2.mac_address, "FE:C4:05:42:8B:AA");
```

Comment:
You can directly compare to a manually created vector here, it will be clearer that you compare the full struct, you may even drop the length check.
```rust
assert_eq!(interfaces, vec![
    Interface{logical_name:"bridge0", mac_address:"FE:..."},
    ...
]);
```